FEATURE: Add `category-list`

carsick opened the pull request:
Displays a list of selected categories in a block

<img width="341" alt="image" src="https://user-images.githubusercontent.com/37538241/214689890-cd6e1bdb-716f-4f5b-a7c6-2995f62c0a2a.png">

<img width="826" alt="image" src="https://user-images.githubusercontent.com/37538241/214689650-12f6bb5b-dad6-44f3-b818-bd19f93addce.png">

<img width="1251" alt="image" src="https://user-images.githubusercontent.com/37538241/214689829-4d814d73-ce98-496c-a7f9-4ac6cec067c3.png">


---

Debating whether to use the category selection box (which feels disconnected from the JSON configuration panel) or manually input the IDs which would allow for greater flexibility (multiple category boxes)...

<img width="1257" alt="image" src="https://user-images.githubusercontent.com/37538241/214697318-84ee1eb9-ea23-4bd6-9a7e-2f2a4566057e.png">

carsick commented:
Sorry, I'm not sure why all my old commits are showing up again, I'm trying to fix this 🙃

pmusaraj commented on javascripts/discourse/components/category-list.js:
```diff
@@ -0,0 +1,15 @@
+import Category from "discourse/models/category";
+import Component from "@glimmer/component";
```
```suggestion
import Component from "@glimmer/component";
import I18n from "I18n";
```

The linter error below is legitimate, we need to import the `I18n` module, even though it works without the import.

pmusaraj commented on javascripts/discourse/templates/components/category-list.hbs:
```diff
@@ -0,0 +1,13 @@
+</h3>
+
+<div class="category-list__container">
+
+  {{#each categoryList as |category|}}
```
```suggestion
  {{#each categoryList as |category|}}
```

Nitpick.

pmusaraj commented on javascripts/discourse/templates/components/category-list.hbs:
```diff
@@ -0,0 +1,13 @@
+      {{category-link category}}
+    </div>
+  {{/each}}
+
+</div>
```
```suggestion
</div>
```

And another one.

pmusaraj commented on javascripts/discourse/components/category-list.js:
```diff
@@ -0,0 +1,15 @@
+    this.blockTitle =
+      this.args?.params?.title || I18n.t(themePrefix("category_list.heading"));
+
+    this.categoryList = settings.category_list.split("|");
```
This component works a little differently from other components, in that each "block" gets its parameters from the same theme setting. 

Is it possible to avoid adding the `category_list` theme setting and have the list of categories be set in the JSON schema setting? We'd have to decide whether to use category IDs or category slugs in there, the interface would not be as friendly, but it would fit better with the data structure of the other blocks in the component. (For example, in `category-topics` we pass the category ID as a parameter.)

pmusaraj commented:
> Debating whether to use the category selection box (which feels disconnected from the JSON configuration panel) or manually input the IDs which would allow for greater flexibility (multiple category boxes)...

Ah yeah, I missed this when I wrote the note above. I'd go for the JSON config here, a little less user-friendly, but more flexible (can have multiple category groupings) and better logical fit with the structure of this component.

carsick commented:
> > Debating whether to use the category selection box (which feels disconnected from the JSON configuration panel) or manually input the IDs which would allow for greater flexibility (multiple category boxes)...
> 
> Ah yeah, I missed this when I wrote the note above. I'd go for the JSON config here, a little less user-friendly, but more flexible (can have multiple category groupings) and better logical fit with the structure of this component.

Updated to use `id` comma-separate values within the JSON block and removed the category selector from the settings!

pmusaraj reviewed (approved):
Looks great, thanks for the updates @carsick. 

Any chance you can add a qunit test? It sometimes helps to catch issues. There are already some tests in this repo you can pull inspiration from, and in your local instance you can run the theme tests at `http://localhost:4200/theme-qunit`.

pmusaraj commented on javascripts/discourse/components/category-list.js:
```diff
@@ -0,0 +1,16 @@
+      this.args?.params?.title || I18n.t(themePrefix("category_list.heading"));
+
+    this.categoryList = this.args?.params?.id.split(",");
+    this.categoryList.forEach(function (id, i, arr) {
```
Might be good to add a quick escape hatch here, something like: 

```
if !this.categoryList {
  return;
}
```

carsick commented:
> Looks great, thanks for the updates @carsick.
> 
> Any chance you can add a qunit test? It sometimes helps to catch issues. There are already some tests in this repo you can pull inspiration from, and in your local instance you can run the theme tests at `http://localhost:4200/theme-qunit`.

Should I be regularly adding these to component PRs?

<img width="718" alt="image" src="https://user-images.githubusercontent.com/37538241/215217794-2accda55-20bc-4d7a-b64e-558fd1adbd76.png">

pmusaraj commented on javascripts/discourse/components/category-list.js:
```diff
@@ -0,0 +1,19 @@
+    this.categoryList.forEach(function (id, i, arr) {
+      arr[i] = Category.findById(id);
+    });
+    if (!this.categoryList) {
```
This needs to be before the `forEach` loop, essentially it protects from the loop failing if `this.categoryList` is empty.

carsick replied:
Ah okay, thank you for the explanation. Does this need to be done for [tag-topics.js](https://github.com/discourse/discourse-right-sidebar-blocks/blob/main/javascripts/discourse/components/tag-topics.js) and [top-topics.js](https://github.com/discourse/discourse-right-sidebar-blocks/blob/main/javascripts/discourse/components/top-topics.js) as well?
